### PR TITLE
[FIX] core: optimize prefetching on large recordsets

### PR DESCRIPTION
**Pull request by rco-odoo**
This patch optimizes the performance of prefetching when iterating on
large recordsets.  The optimization is *transparent*, i.e., it requires
no code change for it to apply.

The worst-case scenario for prefetching is the following: the ORM has to
fetch a field for a given `record`, `record._prefetch_ids` (its prefetch
set) is *very* large (many thousands), and most records in the prefetch
set are already in cache.  This requires the ORM to iterate a lot on the
prefetch set in order to make a batch of records not having the field in
cache.

    records = model.browse(ids)     # large recordset
    for record in records:
        record.foo                  # fetch 'foo' every 1k records

When running such a loop on an empty cache, the overhead of prefetching
(determine a batch) grows as the loop progresses.  The time complexity
of this loop is actually O(N²)...

The overhead of prefetching is minimal when `record._prefetch_ids` is
about the size of a prefetching unit, i.e., 1k records.  This commit
modifies the iterator method such that every record returned by the
iterator has a prefetch set of maximum 1k records.

We measured the time taken by the loop above on an empty cache, before
and after this commit, on a simple model (res.partner.category) with
100k records.  The third measure is a reference one: a `_read` on all
prefetched fields (to fill in the cache) followed by the loop.

                            Total time      Time per 1k records
    Before this commit      3.690s          30ms - 45ms
    After this commit       1.176s          12ms
    Read then loop          1.161s          -

The measures are enlightening: the overhead of prefetching was more than
200% of the reference time, and the time to prefetch records grows as
the iteration goes on!  This commit reduces the overhead of prefetching
to less than 2% of the reference time, and make it scale gracefully with
data size.

### CODE REVIEW

File: odoo/models.py [hunk 1]
```diff
@@ -5550,8 +5550,13 @@ def __len__(self):
 
     def __iter__(self):
         """ Return an iterator over ``self``. """
-        for id in self._ids:
-            yield self._browse(self.env, (id,), self._prefetch_ids)
+        if len(self._ids) > PREFETCH_MAX and self._prefetch_ids is self._ids:
```
Comment (Elkasitu) [inline odoo/models.py:5553]: just a small question: why does this apply only if the prefetch set is the same as the recordset? is there an actual technical reason or just an use case not worth supporting?

**Review by Elkasitu — APPROVED** [review]

File: odoo/models.py (hunk 1 above)
Comment (rco-odoo) [reply odoo/models.py:5553]: It's all about having **predictable performance**.

Consider the following code:
```python
records = model.search(...)
for record in records:              # optimization applies here
    print(record.name)
    for line in record.line_ids:    # optimization does not apply here
        print(line.name)
```
The optimization applies to all recordsets returned by `browse()`.  This includes the result of `search()`, `create()`, `self` in compute methods, `self` in RPC-called methods, etc.  So, in the example above, `record._prefetch_ids` is guaranteed to be no more than `PREFETCH_MAX` ids, which reduces the overhead of the prefetching of `record.name` to the minimum.

However, the situation is quite different for *secondary* recordsets, like `record.line_ids`.  The value of `record.line_ids._prefetch_ids` is an *iterable* object that represents the union of the values of `line_ids` on `record._prefetch_ids` in cache, and the iteration on that object is much more costly!  Given that the arity of the relation `line_ids` is unknown, the cost of iterating on `record.line_ids._prefetch_ids` is essentially not predictable, and slicing its result is tricky, as the size of `record.line_ids` may go from zero to several batches.  Moreover, at the time of `__iter__`, the ORM doesn't even know whether `line._prefetch_ids` will actually be used.

So, for `line.name`, the ORM will iterate on `line._prefetch_ids = record.line_ids._prefetch_ids` to determine prefetching candidates.  That case has potentially more overhead, but the top-level optimization (iteration on `records`) keeps it under relative control.  Indeed, the size of `record.line_ids._prefetch_ids` is around `average_arity_of_line_ids * PREFETCH_MAX`.

Long story short, some nested iterations *may* require some manual optimization, but those cases are much less common than simple loops.

Comment (Elkasitu) [reply odoo/models.py:5553]: Thanks for the explanation, I guess it wasn't such a small question after all :rofl: at least it's documented in the PR now!

**Comment by rco-odoo** [comment]
@robodoo r+
